doc/gengen.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.46.2.
.TH GENGEN "1" "June 2015" "gengen 1.4.3" "User Commands"
.SH NAME
gengen \- manual page for gengen 1.4.3
.SH SYNOPSIS
.B gengen
[\fI\,OPTIONS\/\fR]...
.SH DESCRIPTION
GNU gengen 1.4.3
.PP
This program, given a skeleton source file with some
parts to be replaced, generates a program that generates that source,
and substitutes those parts as specified.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit
.TP
\fB\-i\fR, \fB\-\-input\fR=\fI\,file\/\fR
input file (default std input)
.TP
\fB\-F\fR, \fB\-\-file\-name\fR=\fI\,name\/\fR
name used for generating file (default stdout)
.TP
\fB\-f\fR, \fB\-\-gen\-name\fR=\fI\,STRING\/\fR
prefix for generated code  (default=`gengen')
.TP
\fB\-\-output\-format\fR=\fI\,lang\/\fR
target language (cpp, c, lua)  (default=`cpp')
.TP
\fB\-\-output\-dir\fR=\fI\,STRING\/\fR
output directory (default current dir)
.TP
\fB\-\-separate\-files\fR
output to separate files (header, source). require
\fB\-\-file\-name\fR
.TP
\fB\-\-expand\-tabs\fR
strings containing newlines are indented according
to the current indentation level
.TP
\fB\-\-no\-gen\-version\fR
do not put gengen version in the generated file
.TP
\fB\-\-test\-mode\fR
even with errors exit with 0 (this is only for
testing purposes)
.TP
\fB\-\-force\fR
force regeneration of output code
.SH "REPORTING BUGS"
Report bugs to Raman Gopalan <ramangopalan@gmail.com>
.SH COPYRIGHT
Copyright \(co 2002\-2015  Lorenzo Bettini <http://www.lorenzobettini.it>
This program comes with ABSOLUTELY NO WARRANTY; for details
please see the file 'COPYING' supplied with the source code.
.br
This is free software, and you are welcome to redistribute it
under certain conditions; again, see 'COPYING' for details.
This program is released under the GNU General Public License.
.SH "SEE ALSO"
The full documentation for
.B gengen
is maintained as a Texinfo manual.  If the
.B info
and
.B gengen
programs are properly installed at your site, the command
.IP
.B info gengen
.PP
should give you access to the complete manual.


NEWS
GNU gengen -- History of visible changes.

Copyright (C) 2002-2015  Lorenzo Bettini <www.lorenzobettini.it>

Version 1.4.3

* Implement Lua module for GNU Gengen

Version 1.4.2

* correctly handle --output-dir

Version 1.4.1

* fixed some compilation errors

Version 1.4

* remove possible path information passed with --file-name
* --output-dir option added

Version 1.3

* updated gnulib configure macros
* use --docdir if specified for installing documentation

Version 1.2

* better checks for getopt_long and strdup (using gnulib)
* removed non-standard make rules
* updated to version 3 of GPL

Version 1.1

* distribution of cmdline.ggo whose absence caused make failures in
  doc directory (thanks to Bob White)

Version 1.0

* First GNU release
* conditional statements expressions can be complex
  (not single variables)
* variables can also have types bool and int

Version 0.6

* conditional statements in template files
* the sequence @x@@y@ is now recognized correctly
  as two variables
* before actually generating the file, check whether
  it is different from a possible previous generation

Version 0.5.1

* removed memory leaks in generated C files

Version 0.5

* C code can be generated instead of C++ code
* if an item is of type "method" it can still be assigned a string
and in that case the method is not invoked (the same holds for C external
functions)
* correctly reset the indent counter to 0 when it encounters a new
line and no spaces at the beginning of the line
* --expand-tabs option added that allows to correctly indent
strings with newlines inside.

Version 0.4.4

* fixed some compilation problems under SunOS (thanks to
Dinesh Shanbhag <dshanbhag@hotmail.com>)
* presence of getopt_long is handled in a cleaner way

Version 0.4.3

* in the generated code config.h is not included anymore
(it's useless)

Version 0.4.2

* generate code compliant to new C++ standard headers: removed warnings
with gcc 3.2

Version 0.4.1

* if separated files are generated, no default value (= 0) for indent
is generated (it is already specified in the header file).  This caused
some problems with some gcc 3.x.

src/gengen.cc
/*
 * Copyright (C) 2002-2015  Lorenzo Bettini <http://www.lorenzobettini.it>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <iostream>
#include <string.h>
#include <stdlib.h>

#include "skelstruct.hpp"
#include "skelfactory_cpp.h"
#include "skelfactory_c.h"
#include "skelfactory_lua.h"
#include "mainskelstruct.h"
#include "mainskelfactory.h"
#include "cmdline.h"
#include "fileutil.h"
#include "sourcename.h"
#include "copyright.h"
#include "gengen_cmdline_options.h"
#include "streamstruct.h"
#include "my_sstream.h"

extern int yyparse() ;

SkelStruct *skel_struct = 0;
SkelItems *skel_items = 0;
SkelFactory *skel_factory = 0;

string input_file;
string stripped_input_file;

using namespace std;

static void
gengen_error(const string &s);

static void
generate_code(const ostringstream &stream, const string &file_name);

gengetopt_args_info args_info;

int
main(int argc, char **argv)
{
  if(cmdline_parser(argc, argv, &args_info) != 0)
    // calls cmdline parser. The user gived bag args if it doesn't return -1
    return 1;

  if (args_info.help_given)
  {
    cout << "GNU ";
    cmdline_parser_print_help ();
    cout << endl;
    cout << "Report bugs to Raman Gopalan <ramangopalan@gmail.com> " << endl;
    exit (0);
  }

  if (args_info.version_given)
  {
    cout << "GNU ";
    cmdline_parser_print_version ();
    copyright_gen_class copyright;
    copyright.set_year ("2002-2015");
    copyright.generate_copyright (cout);
    exit (0);
  }

  string gen_name;
  string file_name;

  gen_name = string (args_info.gen_name_arg);

  expand_tabs = (args_info.expand_tabs_given != 0);
  no_gen_version = (args_info.no_gen_version_given != 0);

  if (args_info.test_mode_given)
    test_mode = true;

  if (args_info.file_name_given)
    file_name = string (args_info.file_name_arg);

  if (args_info.output_dir_given)
	  file_name = string(args_info.output_dir_arg) + "/" + strip_file_path(file_name);

  if (args_info.input_given)
    {
      input_file = string (args_info.input_arg);
      stripped_input_file = strip_file_path(input_file);
      open_input_file (input_file);
      // otherwise stdin already points to stdin
    }

  string source_ext;
  if (strcmp (args_info.output_format_arg, "cpp") == 0)
    {
      skel_factory = new SkelFactoryCpp;
      source_ext = "cc";
    }
  else if (strcmp (args_info.output_format_arg, "c") == 0)
    {
      skel_factory = new SkelFactoryC;
      source_ext = "c";
    }
  else if (strcmp (args_info.output_format_arg, "lua") == 0)
    {
      skel_factory = new SkelFactoryLua;
      source_ext = "lua";
    }
  else
    {
      string serr = "target language ";
      serr += args_info.output_format_arg;
      serr += " not recognized";
      gengen_error (serr);
    }

  if (args_info.separate_files_given && ! args_info.file_name_given)
    {
      gengen_error ("--separate-files requires --file-name option");
    }

  if (yyparse () == EXIT_FAILURE)
    return EXIT_FAILURE;

  StreamStruct streamstruct;

  ostringstream header_s;
  ostringstream source_s;
  if (args_info.file_name_given)
    {
      streamstruct.header_stream = &header_s;
      if (args_info.separate_files_given)
        streamstruct.source_stream = &source_s;
    }
  else
    streamstruct.header_stream = &cout;

  skel_struct = skel_factory->createSkelStruct
       (skel_items, gen_name, file_name, streamstruct,
         args_info.separate_files_given);

  skel_struct->generate ();

  if (args_info.file_name_given)
    {
      generate_code (header_s, create_output_filename(file_name));
      if (args_info.separate_files_given)
        {
          generate_code(source_s, create_output_filename(file_name, source_ext));
        }
    }

  return 0;
}

// check whether any change took place since the last generated file (if any exists)
void
generate_code(const ostringstream &stream, const string &file_name)
{
  const istream *previous = open_input_file_stream(file_name);
  bool generate = true;

  cerr << "generating " << file_name << "... ";

  const string &stream_buffer = stream.str();

  if ((! args_info.force_given) && previous && *previous)
    {
      ostringstream previous_s;
      previous_s << previous->rdbuf();
      if (previous_s.str() == stream_buffer)
        generate = false;
    }

  if (generate)
    {
      ostream &output_file = open_output_file(file_name);
      output_file << stream_buffer << flush;
      cerr << "done" << endl;
    }
  else
    cerr << "no change needed" << endl;
}

void
gengen_error(const string &s)
{
  cerr << "gengen: " << s << endl;
  exit (EXIT_FAILURE);
}
